concordia/admin/views.py

```diff
@@ -10,11 +10,100 @@
 from django.views.decorators.cache import never_cache
 from tabular_export.core import export_to_csv_response, flatten_queryset
 
-from importer.tasks import import_items_into_project_from_url
+from importer.tasks import import_items_into_project_from_url, redownload_image_task
 from importer.utils.excel import slurp_excel
 
-from ..models import Campaign, Project, SiteReport
-from .forms import AdminProjectBulkImportForm
+from ..models import Asset, Campaign, Project, SiteReport
+from .forms import AdminProjectBulkImportForm, AdminRedownloadImagesForm
+
+
+@never_cache
+@staff_member_required
+@permission_required("concordia.add_item")
+@permission_required("concordia.change_item")
+def redownload_images_view(request):
+    request.current_app = "admin"
+
+    context = {"title": "Redownload Images"}
+
+    if request.method == "POST":
+        form = AdminRedownloadImagesForm(request.POST, request.FILES)
+
+        if form.is_valid():
+            context["assets_to_download"] = assets_to_download = []
+
+            rows = slurp_excel(request.FILES["spreadsheet_file"])
+            required_fields = [
+                "download_url",
+            ]
+            for idx, row in enumerate(rows):
+                missing_fields = [i for i in required_fields if i not in row]
+                if missing_fields:
+                    messages.warning(
+                        request, f"Skipping row {idx}: missing fields {missing_fields}"
+                    )
+                    continue
+
+                download_url = row["download_url"]
+                # optional real_file_url data
+                real_file_url = row["real_file_url"]
+
+                if not download_url:
+                    if not any(row.values()):
+                        # No messages for completely blank rows
+                        continue
+
+                    warning_message = (
+                        f"Skipping row {idx}: the required field "
+                        "download_url is empty"
+                    )
+                    messages.warning(request, warning_message)
+                    continue
+
+                if not download_url.startswith("http"):
+                    messages.warning(
+                        request, f"Skipping unrecognized URL value: {download_url}"
+                    )
+                    continue
+
+                try:
+                    # Use the download_url to look up the related asset.
+                    # Then queue the task to redownload the image file.
+                    assets = Asset.objects.filter(download_url=download_url)
```

Comment:
Can this ever match more than one `Asset`? I'm wondering whether this should just be an `Asset.objects.get()` with an exception handler for the `DoesNotExist` path.

Reply:
Yes and it does in our current database. Not sure why.

Reply:
If the asset does not exist (in our current situation that's about 40% of the time) the "No matching asset" message is displayed in the messages output.